drill.cpp

```diff
@@ -383,32 +383,48 @@ bool ExcellonProcessor::millhole(std::ofstream &of, double start_x, double start
     // cutter->stepsize in depth.
-    unsigned int stepcount = 1;
+    int stepcount = 1;
     if (cutter->do_steps) {
-        stepcount = (unsigned int) ceil(abs(cutter->zwork / cutter->stepsize));
+        stepcount = (int) ceil(abs(cutter->zwork / cutter->stepsize));
     }
 
+    double delta_x = stop_x - start_x;
+    double delta_y = stop_y - start_y;
+    double distance = sqrt(delta_x*delta_x + delta_y*delta_y);
     if (cutdiameter * 1.001 >= holediameter) { //In order to avoid a "zero radius arc" error
         // Hole is smaller than cutdiameter so just drill/zig-zag.
         of << "G0 X" << start_x * cfactor << " Y" << start_y * cfactor << '\n';
-        if (slot)
-        {
-            for (unsigned int current_step = 0; true;)
-            {
-                double z = double(current_step+1)/(stepcount) * cutter->zwork;
-                of << "G1 Z" << z * cfactor << '\n';
-                of << "G1 X" << stop_x * cfactor << " Y" << stop_y * cfactor << '\n';
-                current_step++;
-                if (current_step >= stepcount) {
-                    break;
+        if (slot) {
+            double zhalfstep = cutter->zwork / stepcount / 2;
+            double feedrate;
+            if (holediameter + distance < 1.3 * cutdiameter) {
+                // This is mostly a plunge.
+                feedrate = cutter->vertfeed;
+            } else {
+                // This is mostly a zig-zag.
+                feedrate = cutter->feed;
+            }
+            // Start one step above Z0 for optimal entry
+            of << "G1 Z" << -1.0/stepcount * cutter->zwork * cfactor
+              << " F" << feedrate * cfactor << '\n';
+            for (int current_step = -1; true; current_step++) {
+                // current_step == stepcount is for the bottom pass, so z needs to stay the same
+                double z = double(std::min(stepcount, current_step+1))/stepcount * cutter->zwork;
+                of << "G1 X" << stop_x * cfactor
+                   << " Y" << stop_y * cfactor;
+                if(stepcount != current_step) {
+                   // Drop superfluous Z from the bottom pass to indicate that this line is not there by accident
+                   of << " Z" << (z - zhalfstep) * cfactor;
                 }
-                z = double(current_step+1)/(stepcount) * cutter->zwork;
-                of << "G1 Z" << z * cfactor << '\n';
-                of << "G1 X" << start_x * cfactor << " Y" << start_y * cfactor << '\n';
-                current_step++;
+                of << '\n';
+                // We don't need a second "zag" on the bottom pass
                 if (current_step >= stepcount) {
                     break;
                 }
+                of << "G1 X" << start_x * cfactor
+                   << " Y" << start_y * cfactor
+                   << " Z" << z * cfactor
+                   << '\n';
             }
         } else {
             of << "G1 Z" << cutter->zwork * cfactor << '\n';
```

Comment:
Now that we're messing around with the Feedrate so often, maybe we need to be explicit here.  Because we can't count on the most recent feedrate to be for a plunge.  Am I right?

Reply:
Oh. My mistake. I thought it was just always the same, but I removed one. I will fix this.